From 7ef2ce4f90b6f2b41d35cd8301478c880dd81926 Mon Sep 17 00:00:00 2001
From: Julian Smith <julian.smith@artifex.com>
Date: Wed, 24 Jun 2026 05:50:52 +0100
Subject: [PATCH] scripts/wrap/cpp.py: fix pod operator== for array members.

We need to use memcmp() if a struct member is an array.
---
 scripts/wrap/cpp.py | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/scripts/wrap/cpp.py b/scripts/wrap/cpp.py
index c357d50a33..c1c62bae47 100644
--- a/scripts/wrap/cpp.py
+++ b/scripts/wrap/cpp.py
@@ -4209,7 +4209,11 @@ def pod_struct_fns(
     out_cpp.write( f'FZ_FUNCTION bool operator==( const ::{struct_name}& lhs, const ::{struct_name}& rhs)\n')
     out_cpp.write( f'{{\n')
     for cursor in parse.get_members(struct_cursor):
-        out_cpp.write( f'    if (lhs.{cursor.spelling} != rhs.{cursor.spelling}) return false;\n')
+        if cursor.type.get_array_size() >= 0:
+            # This struct member is an array, so compare with memcmp().
+            out_cpp.write( f'    if (memcmp(lhs.{cursor.spelling}, rhs.{cursor.spelling}, sizeof(lhs.{cursor.spelling}))) return false;\n')
+        else:
+            out_cpp.write( f'    if (lhs.{cursor.spelling} != rhs.{cursor.spelling}) return false;\n')
     out_cpp.write( f'    return true;\n')
     out_cpp.write( f'}}\n')
     out_cpp.write( f'FZ_FUNCTION bool operator!=( const ::{struct_name}& lhs, const ::{struct_name}& rhs)\n')
